src/dataflow/decode/mod.rs

```diff
@@ -118,6 +118,306 @@ where
     })
 }
 
+pub type PushSession<'a, R> =
+    Session<'a, Timestamp, R, PushCounter<Timestamp, R, Tee<Timestamp, R>>>;
+
+pub trait DecoderState {
+    /// Reset number of success and failures with decoding
+    fn reset_event_count(&mut self);
+    fn decode_key(&mut self, bytes: &[u8]) -> Result<Row, String>;
+    /// give a session a key-value pair
+    fn give_key_value<'a>(
+        &mut self,
+        key: Row,
+        bytes: &[u8],
+        aux_num: &Option<i64>,
```

Comment:
There is no reason for this to be a reference, since `Option<i64>` is `Copy`, it can be passed by value.